common/djangoapps/student/models.py

```diff
@@ -166,13 +169,30 @@ def anonymous_id_for_user(user, course_id, save=True):
     if cached_id is not None:
         return cached_id
 
-    # include the secret key as a salt, and to make the ids unique across different LMS installs.
-    hasher = hashlib.md5()
+    # check if an annonymous id already exists for this user and course_id combination
+    anonymous_user_ids = AnonymousUserId.objects.filter(user=user).filter(course_id=course_id)
+    if anonymous_user_ids:
+        if len(anonymous_user_ids) == 1:
+            return getattr(anonymous_user_ids[0], "anonymous_user_id")
+        else:
+            # there should only be one anonymous_user_id per user/course_id pair
+            raise Exception('Expected only one anonymous_user_id for user/course_id pair, instead ther are {}'.format(len(anonymous_user_ids)))
+
+    # include the ANONYMOUS_ID_PEPPER as salt/pepper, and to make the ids unique across different LMS installs.
+    hasher = hashlib.shake_256()
+    # This is one of several uses of SECRET_KEY.
+    #
+    # Impact of exposure: Could result in PII exposure
```

Comment:
```suggestion
    # Impact of exposure: If a person has the SECRET_KEY and a user's `id` they can correlate the users anonymous user IDs across any courses they have participated in.
```
I think something more specific would be good here.